File: CustomObjects/vdPolygon/vdLineConnection.cs
using System;
using System.Collections.Generic;
using System.Text;
using VectorDraw.Geometry;
using VectorDraw.Professional.vdCollections;
using VectorDraw.Professional.vdFigures;
using VectorDraw.Professional.vdObjects;
using VectorDraw.Professional.vdPrimaries;
using VectorDraw.Render;
using VectorDraw.Serialize;

namespace VectorDrawCustomFigs
{
    public class vdLineConnection : vdFigure, IvdProxyFigure, ISupportResolveReferenceObjects
    {
        vdFigure mFig1 = null;
        vdFigure mFig2 = null;

        public vdLineConnection()
        {
        }
        public vdLineConnection(vdDocument doc) 
        {
            SetUnRegisterDocument(doc);
            setDocumentDefaults();
        }

        public vdFigure Figure1
        {
            get { return mFig1; }
            set
            {
                AddHistory("Figure1", value);
                Disconnect(mFig1);
                mFig1 = value;
                Connect(mFig1);
            }
        }
        public vdFigure Figure2
        {
            get { return mFig2; }
            set
            {
                AddHistory("Figure2", value);
                Disconnect(mFig2);
                mFig2 = value;
                Connect(mFig2);
            }
        }
        bool Isvalid
        {
            get
            {
                if (Figure1 == null || Figure2 == null || Figure1 == Figure2 || Figure1.Deleted || Figure2.Deleted) return false;
                return true;
            }
        }
        public override bool Deleted
        {
            get
            {
                if (!Isvalid) return true;
               return base.Deleted;
            }
        }
        public override Box BoundingBox
        {
            get
            {
                if (mBoundBox.IsEmpty)
                {
                    if (Isvalid)
                    {
                        //mBoundBox = new Box(Figure1.BoundingBox.MidPoint, Figure2.BoundingBox.MidPoint);
                        mBoundBox.AddBox(Figure1.BoundingBox);
                        mBoundBox.AddBox(Figure2.BoundingBox);
                    }
                }
                return mBoundBox;
            }

        }
        public override vdRender.DrawStatus Draw(vdRender render)
        {
            if (!Isvalid) return vdRender.DrawStatus.Failed;
            vdRender.DrawStatus doDraw = base.Draw(render);
            if (doDraw == vdRender.DrawStatus.Successed)
            {
                render.DrawLine(this, Figure1.BoundingBox.MidPoint, Figure2.BoundingBox.MidPoint);
            }
            AfterDraw(render);
            return doDraw;
        }
        public override void Serialize(Serializer serializer)
        {
            base.Serialize(serializer);
            serializer.Serialize(Figure1, "Figure1");
            serializer.Serialize(Figure2, "Figure2");
        }
        public override gPoints GetGripPoints()
        {
            gPoints ret = new gPoints();
            if (Figure1 != null && Figure2 != null) ret.Add(gPoint.MidPoint(Figure1.BoundingBox.MidPoint, Figure2.BoundingBox.MidPoint));
            return ret;
        }
        public override void MoveGripPointsAt(Int32Array Indexes, double dx, double dy, double dz)
        {
            if (Indexes == null || Indexes.Count == 0 || (dx == 0.0 && dy == 0.0 && dz == 0.0)) return;
            gPoints grips = GetGripPoints();

            Matrix matt = new Matrix();
            matt.TranslateMatrix(new gPoint(dx, dy, dz));
            foreach (int index in Indexes)
            {
                switch (index)
                {
                    case 0:
                        if (Figure1 != null) Figure1.Transformby(matt);
                        if (Figure2 != null) Figure2.Transformby(matt);
                        break;
                    default:
                        break;
                }
            }
            Update();
        }
        public override bool DeSerialize(DeSerializer deserializer, string fieldname, object value)
        {
            if (base.DeSerialize(deserializer, fieldname, value)) return true;
            else if (fieldname == "Figure1") Figure1 = (vdFigure)value;
            else if (fieldname == "Figure2") Figure2 = (vdFigure)value;
            else return false;
            return true;
        }

        void ISupportResolveReferenceObjects.ResolveReferences(vdFigure sourcefigure, ResolveReferenceObjects referenceObjects)
        {
            vdLineConnection baseobj = sourcefigure as vdLineConnection;
            Figure1 = referenceObjects.AddReference(baseobj.Figure1);
            Figure2 = referenceObjects.AddReference(baseobj.Figure2);
        }

        void Connect(vdFigure fig)
        {
            if (fig == null) return;
            fig.OnAfterModifyObjectEx += Fig_OnAfterModifyObjectEx;
        }
        void Disconnect(vdFigure fig)
        {
            if (fig == null) return;
            fig.OnAfterModifyObjectEx -= Fig_OnAfterModifyObjectEx;
        }

        private void Fig_OnAfterModifyObjectEx(object sender, string propertyname)
        {
            this.Update();
            this.Invalidate();
        }

        protected override void OnOwnerChanged()
        {
            base.OnOwnerChanged();
            if(Owner == null)
            {
                Disconnect(mFig1);
                Disconnect(mFig2);
            }
            
        }
        void ISupportResolveReferenceObjects.OnAddToSameSelection(vdSelection selset)
        {
        }
        private void VdLineConnection_OnAfterModifyObjectEx(object sender, string propertyname)
        {
            this.Update();
            this.Invalidate();
        }
        /// <summary>
        /// A simple command that will prompt the user to select two vdFigures and then create a LineConnection between them.
        /// </summary>
        /// <param name="doc"></param>
        public static void LineConnectionEx(vdDocument doc)
        {
             vdFigure fig1 = null, fig2 = null;
            gPoint pt;
            doc.Prompt("Select Fig1");
            doc.ActionUtility.getUserEntity(out fig1, out pt);
            doc.Prompt(null);
            if (fig1 != null)
            {
                doc.Prompt("Select Fig2");
                doc.ActionUtility.getUserEntity(out fig2, out pt);
                doc.Prompt(null);
            }

            if(fig1 == null || fig2 == null || fig1 == fig2)
            {
                doc.Prompt("\r\nYou must select 2 deferent entities:");
                doc.Prompt(null);
            }
            vdLineConnection lineconnection = new vdLineConnection(doc);
            lineconnection.Figure1 = fig1;
            lineconnection.Figure2 = fig2;
            doc.Model.Entities.AddItem(lineconnection);
            doc.ActionDrawFigure(lineconnection);
        }
    }
}

File: VB/Customfigs_VB/VectorDrawCustomFigs_VB/vdLineConnection.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports VectorDraw.Geometry
Imports VectorDraw.Professional.vdCollections
Imports VectorDraw.Professional.vdFigures
Imports VectorDraw.Professional.vdObjects
Imports VectorDraw.Professional.vdPrimaries
Imports VectorDraw.Render
Imports VectorDraw.Serialize

Public Class vdLineConnection
    Inherits vdFigure
    Implements IvdProxyFigure, ISupportResolveReferenceObjects

#Region "Private fields"
    Private mFig1 As vdFigure
    Private mFig2 As vdFigure
#End Region

#Region "ctor"
    Public Sub New()
    End Sub
    Public Sub New(doc as vdDocument)
        SetUnRegisterDocument(doc)
        setDocumentDefaults()
    End Sub
#End Region

#Region "Public properties"
    Public Property Figure1() As vdFigure
        Get
            Return mFig1
        End Get
        Set(ByVal Value As vdFigure)
            AddHistory("Figure1", Value)
            Disconnect(mFig1)
            mFig1 = Value
            Connect(mFig1)
        End Set
    End Property
    Public Property Figure2() As vdFigure
        Get
            Return mFig2
        End Get
        Set(ByVal Value As vdFigure)
            AddHistory("Figure2", Value)
            Disconnect(mFig2)
            mFig2 = Value
            Connect(mFig2)
        End Set
    End Property
    Private ReadOnly Property Isvalid() As Boolean
        Get
            If ((Figure1 Is Nothing) Or (Figure2 Is Nothing)) Then Return False
            If ((Figure1.Equals(Figure2)) Or (Figure1.Deleted) Or (Figure2.Deleted)) Then Return False
            Return True
        End Get
    End Property
    Public Overrides Property Deleted() As Boolean
        Get
            If (Not Isvalid) Then Return True
            Return MyBase.Deleted
        End Get
        Set(ByVal value As Boolean)
            MyBase.Deleted = value
        End Set
    End Property
    Public Overrides ReadOnly Property BoundingBox() As Box
        Get

            If (mBoundBox.IsEmpty) Then
                If (Isvalid) Then
                    mBoundBox.AddBox(Figure1.BoundingBox)
                    mBoundBox.AddBox(Figure2.BoundingBox)
                End If
            End If
            Return mBoundBox
        End Get
    End Property
    

#End Region

#Region "Methods"
    Public Overrides Function Draw(ByVal render As vdRender) As vdRender.DrawStatus

        If (Not Isvalid) Then Return vdRender.DrawStatus.Failed
        Dim doDraw As vdRender.DrawStatus = MyBase.Draw(render)
        If doDraw = vdRender.DrawStatus.Successed Then
            render.DrawLine(Me, Figure1.BoundingBox.MidPoint, Figure2.BoundingBox.MidPoint)
        End If
        AfterDraw(render)
        Return doDraw
    End Function
    Public Overrides Sub Serialize(ByVal serializer As Serializer)
        MyBase.Serialize(serializer)
        serializer.Serialize(Figure1, "Figure1")
        serializer.Serialize(Figure2, "Figure2")
    End Sub
    Public Overrides Function DeSerialize(ByVal deserializer As DeSerializer, ByVal fieldname As String, ByVal value As Object) As Boolean
        If MyBase.DeSerialize(deserializer, fieldname, value) Then
            Return True
        ElseIf (fieldname = "Figure1") Then
            Figure1 = CType(value, vdFigure)
        ElseIf (fieldname = "Figure2") Then
            Figure2 = CType(value, vdFigure)
        Else
            Return False
        End If
        Return True
    End Function
    Public Overrides Function GetGripPoints() As gPoints
        Dim ret As gPoints = New gPoints()
        If (Figure1 Is Nothing Or Figure2 Is Nothing) Then Return ret
        ret.Add(gPoint.MidPoint(Figure1.BoundingBox.MidPoint, Figure2.BoundingBox.MidPoint))
        Return ret
    End Function
    Public Overrides Sub MoveGripPointsAt(ByVal Indexes As Int32Array, ByVal dx As Double, ByVal dy As Double, ByVal dz As Double)
        If Indexes Is Nothing Or Indexes.Count = 0 Or (dx = 0.0 And dy = 0.0 And dz = 0.0) Then
            Return
        End If
        Dim grips As gPoints = GetGripPoints()
        Dim mat As Matrix = New Matrix()
        mat.TranslateMatrix(dx, dy, dz)
        Dim index As Integer
        For Each index In Indexes
            Select Case index
                Case 0
                    If (Not Figure1 Is Nothing) Then Figure1.Transformby(mat)
                    If (Not Figure2 Is Nothing) Then Figure2.Transformby(mat)
                    Exit For
                Case Else
                    Exit For
            End Select
        Next
        Update()
    End Sub
    Private Sub Connect(ByVal fig As vdFigure)
        If (fig Is Nothing) Then Return
        AddHandler fig.OnAfterModifyObjectEx, AddressOf Fig_OnAfterModifyObjectEx
    End Sub
    Private Sub Disconnect(ByVal fig As vdFigure)
        If (fig Is Nothing) Then Return
        RemoveHandler fig.OnAfterModifyObjectEx, AddressOf Fig_OnAfterModifyObjectEx
    End Sub
    Private Sub Fig_OnAfterModifyObjectEx(ByVal sender As Object, ByVal propertyname As String)
        Me.Update()
        Me.Invalidate()
    End Sub
    Private Sub ResolveReferences(ByVal sourcefigure As vdFigure, ByVal referenceObjects As ResolveReferenceObjects) Implements ISupportResolveReferenceObjects.ResolveReferences
        Dim baseobj As vdLineConnection = CType(sourcefigure, vdLineConnection)
        Figure1 = referenceObjects.AddReference(baseobj.Figure1)
        Figure2 = referenceObjects.AddReference(baseobj.Figure2)
    End Sub
    Private Sub OnAddToSameSelection(ByVal selset As vdSelection) Implements ISupportResolveReferenceObjects.OnAddToSameSelection
    End Sub
    Protected Overrides Sub OnOwnerChanged()
        MyBase.OnOwnerChanged()
        If (Owner Is Nothing) Then
            Disconnect(mFig1)
            Disconnect(mFig2)
        End If
    End Sub

#End Region
End Class
